src/main/java/li/cil/oc2/common/inet/SessionBase.java

```diff
@@ -0,0 +1,79 @@
+package li.cil.oc2.common.inet;
+
+import li.cil.oc2.api.inet.Session;
+import li.cil.oc2.common.Config;
+
+import javax.annotation.Nullable;
+import java.net.InetSocketAddress;
+import java.time.Instant;
+import java.util.concurrent.atomic.AtomicLong;
+
+public abstract class SessionBase implements Session {
+    private static final AtomicLong idGenerator = new AtomicLong();
+
+    private final long id = idGenerator.getAndIncrement();
+    private final InetSocketAddress destination;
+    private States state;
+    private Instant expireTime;
+    private Object userdata;
+
+    public SessionBase(final int ipAddress, final short port) {
+        destination = new InetSocketAddress(InetUtils.toJavaInetAddress(ipAddress), Short.toUnsignedInt(port));
+        state = States.NEW;
+    }
+
+    @Override
+    public long getId() {
+        return id;
+    }
+
+    @Override
+    public void close() {
+        switch (state) {
+            case NEW:
+                state = States.REJECT;
+                return;
+            case ESTABLISHED:
+                state = States.FINISH;
+                return;
+            default:
+                throw new IllegalStateException();
```

Comment:
I have limited the number of sent and received packets to one packet per tick. I am assuming that any update of the `Session` object can generate a packet. This includes invocation of the `Session::close` method. This will not happen with UDP, but in TCP there will be at least FIN packet. This means that the `SessionLayer` implementation must take into account that calling the close method on the `Session` object should update that object and then call the `SessionLayer.Receiver::receive` method. Because of this, I am assuming that the double update of the `Session` object indicates a bug in the` SessionLayer` implementation.